### PR TITLE
Add deferrable mode to `GCSObjectUpdateSensor`

### PR DESCRIPTION
**Pull request by phanikumv**
This PR donates the `GCSObjectUpdateSensorAsync` from [astronomer-providers](https://github.com/astronomer/astronomer-providers) to Airflow.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/google/cloud/sensors/gcs.py [hunk 1]
```diff
@@ -232,6 +234,7 @@ def __init__(
             )
         self.delegate_to = delegate_to
         self.impersonation_chain = impersonation_chain
+        self.deferrable = (deferrable,)
```
Comment (dstandish) [inline airflow/providers/google/cloud/sensors/gcs.py]: this seems odd?

File: airflow/providers/google/cloud/sensors/gcs.py [hunk 2]
```diff
@@ -242,6 +245,42 @@ def poke(self, context: Context) -> bool:
         )
         return hook.is_updated_after(self.bucket, self.object, self.ts_func(context))
 
+    def execute(self, context: Context) -> None:
+        """Airflow runs this method on the worker and defers using the trigger."""
+        if not self.deferrable:
```
Comment (dstandish) [inline airflow/providers/google/cloud/sensors/gcs.py]: `is True` ?

File: airflow/providers/google/cloud/sensors/gcs.py [hunk 3]
```diff
@@ -242,6 +245,42 @@ def poke(self, context: Context) -> bool:
...
+    def execute_complete(self, context: dict[str, Any], event: dict[str, str] | None = None) -> str:
+        """
+        Callback for when the trigger fires - returns immediately.
+        Relies on trigger to throw an exception, otherwise it assumes execution was
+        successful.
+        """
+        if event:
+            if event["status"] == "success":
+                self.log.info(
+                    "Sensor checks update time for object %s in bucket : %s", self.object, self.bucket
```
Comment (dstandish) [inline airflow/providers/google/cloud/sensors/gcs.py]: not sure the meaning of this message?

File: airflow/providers/google/cloud/triggers/gcs.py [hunk 4]
```diff
@@ -97,3 +99,101 @@ async def _object_exists(self, hook: GCSAsyncHook, bucket_name: str, object_name
...
+    :param poke_interval: polling period in seconds to check for file/folder
+    :param google_cloud_conn_id: reference to the Google Connection
+    :param hook_params: dict object that has delegate_to and impersonation_chain
+    """
+
+    def __init__(
+        self,
+        bucket: str,
+        object_name: str,
+        ts: datetime,
```
Comment (dstandish) [inline airflow/providers/google/cloud/triggers/gcs.py]: this variable name should make at least _some_ effort convey what it usage / meaning / intent is

or, at least not be a 2 letter abbreviation

File: airflow/providers/google/cloud/triggers/gcs.py [hunk 5]
```diff
@@ -97,3 +99,101 @@ async def _object_exists(self, hook: GCSAsyncHook, bucket_name: str, object_name
...
+                "object_name": self.object_name,
+                "ts": self.ts,
+                "poke_interval": self.poke_interval,
+                "google_cloud_conn_id": self.google_cloud_conn_id,
+                "hook_params": self.hook_params,
+            },
+        )
+
+    async def run(self) -> AsyncIterator["TriggerEvent"]:
+        """Simple loop until the object updated time is greater than ts datetime in bucket."""
```
Comment (dstandish) [inline airflow/providers/google/cloud/triggers/gcs.py]:
```suggestion
        """Loop until the object updated time is greater than ts datetime in bucket."""
```

we don't nec. need to assert that this loop is simple :) we can let the reader be the judge of that 😛

File: airflow/providers/google/cloud/triggers/gcs.py [hunk 6]
```diff
@@ -97,3 +99,101 @@ async def _object_exists(self, hook: GCSAsyncHook, bucket_name: str, object_name
...
             return "pending"
+
+
+class GCSCheckBlobUpdateTimeTrigger(BaseTrigger):
+    """
+    A trigger that makes an async call to GCS to check whether the object is updated in a bucket.
+
+    :param bucket: google cloud storage bucket name cloud storage where the objects are residing.
+    :param object_name: the file or folder present in the bucket
+    :param ts: datetime object
```
Comment (dstandish) [inline airflow/providers/google/cloud/triggers/gcs.py]: missing description of meaning of param

File: airflow/providers/google/cloud/sensors/gcs.py (hunk 1 above)
Comment (phanikumv) [reply airflow/providers/google/cloud/sensors/gcs.py]: yeah this is fixed now

File: airflow/providers/google/cloud/sensors/gcs.py (hunk 2 above)
Comment (phanikumv) [reply airflow/providers/google/cloud/sensors/gcs.py]: changed to `if self.deferrable is False`

File: airflow/providers/google/cloud/sensors/gcs.py (hunk 3 above)
Comment (phanikumv) [reply airflow/providers/google/cloud/sensors/gcs.py]: corrected it

File: airflow/providers/google/cloud/triggers/gcs.py (hunk 5 above)
Comment (phanikumv) [reply airflow/providers/google/cloud/triggers/gcs.py]: corrected

File: airflow/providers/google/cloud/triggers/gcs.py (hunk 6 above)
Comment (phanikumv) [reply airflow/providers/google/cloud/triggers/gcs.py]: added missing description of meaning of param

**Review by potiuk — APPROVED** [review]

**Pull request merged**